// stl2image/lib/to/image.js
(function() {
  var Mustache, Stl, child_proc, fs, os, path, stl_pov;

  Mustache = require('mustache');

  stl_pov = require('./povray');

  os = require('os');

  fs = require('fs');

  path = require('path');

  child_proc = require('child_process');

  Stl = (function() {
    function Stl() {}

    return Stl;

  })();

  Stl.Image = (function() {
    function Image() {
      this.filePath = __dirname + "/../../template/povray.tmpl";
    }

    Image.prototype.convertFile = function(filePath) {
      var callback, convertSTL2POV, k, locals, options, progressCb, progresscb, start, stderrStr, thenInsertIntoTemplate, thenRenderToImage, thenWriteToTempFile, v, _ref;
      options = {
        height: 240,
        width: 320,
        phi: Math.PI / 8,
        dst: "./result.png"
      };
      stderrStr = "";
      locals = {};
      if (typeof arguments[1] === 'function') {
        callback = arguments[1];
        progressCb = arguments[2];
      } else {
        _ref = arguments[1];
        for (k in _ref) {
          v = _ref[k];
          options[k] = v;
        }
        callback = arguments[2];
        progresscb = arguments[3];
      }
      start = convertSTL2POV = (function(_this) {
        return function() {
          return stl_pov.convertFile(filePath, function(err, povData, name) {
            if (err != null) {
              callback(err);
              return;
            }
            locals.modelData = povData;
            locals.modelName = name;
            locals.phi = phi;
            locals.fileName = name + '_' + Math.floor(Math.random() * 1000000 + 1);
            return thenInsertIntoTemplate();
          }, function(err, povPolygon, name) {
            if (err != null) {
              callback(err);
              return;
            }
            if (progressCb != null) {
              return progressCb();
            }
          });
        };
      })(this);
      thenInsertIntoTemplate = (function(_this) {
        return function() {
          return _this.renderTemplate(_this.filePath, locals, function(err, povInput) {
            if (err != null) {
              callback(err);
              return;
            }
            return thenWriteToTempFile(povInput);
          });
        };
      })(this);
      thenWriteToTempFile = (function(_this) {
        return function(povInput) {
          return fs.writeFile("/tmp/" + locals.fileName + ".pov", povInput, function(err) {
            if (err != null) {
              callback(err);
              return;
            }
            return thenRenderToImage('temp.pov');
          });
        };
      })(this);
      thenRenderToImage = (function(_this) {
        return function(povFileName) {
          var isWin, povcmd;
          isWin = /^win/.test(os.platform());
          if (isWin) {
            povcmd = child_proc.spawn('pvengine', ['/exit', "/render", "C:\\tmp\\" + locals.fileName + ".pov", "+W" + options.width, "+H" + options.height, "-o" + options.dst, "+Q9", "+AM1", "+A"]);
          } else {
            povcmd = child_proc.spawn('povray', ["-i/tmp/" + locals.fileName + ".pov", "+FN", "+W" + options.width, "+H" + options.height, "-o" + options.dst, "+Q9", "+AM1", "+A", "+UA"]);
          }
          povcmd.stderr.on('data', function(data) {
            return stderrStr += data.toString();
          });
          return povcmd.on('exit', function(code) {
            var parsedError;
            if (code === 0) {
              return callback(null, stderrStr);
            } else {
              parsedError = _this._parseError(stderrStr);
              if (parsedError != null) {
                return callback(parsedError, stderrStr);
              } else {
                return callback(new Error("Exit code: " + code), stderrStr);
              }
            }
          });
        };
      })(this);
      return start();
    };

    Image.prototype.renderTemplate = function(filePath, locals, callback) {
      return fs.readFile(filePath, (function(_this) {
        return function(err, buffer) {
          var result;
          if (err != null) {
            callback(err);
            return;
          }
          result = Mustache.render(buffer.toString(), locals);
          return callback(null, result);
        };
      })(this));
    };

    Image.prototype._parseError = function(stderrstr) {
      var md;
      md = stderrstr.match(/(Integer parameter expected.*)/g);
      if (md !== null) {
        return new Error(md[0]);
      } else {
        return null;
      }
    };

    return Image;

  })();

  module.exports = new Stl.Image();

}).call(this);
